File: python/tvm/ir/memory_pools.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Objects for Memory Pools to be used within the compilation"""

from typing import Optional, List

from tvm._ffi import register_object
from tvm.runtime import Object
from tvm.runtime import NDArray
from . import _ffi_api


@register_object("ir.PoolInfo")
class PoolInfo(Object):
    """PoolInfo object holds information related to memory pools
    where the statically sized allocate nodes will pooled into.
    This is a base class for WorkspacePoolInfo and ConstantPoolInfo.
    """

    def __init__(self):
        pass


@register_object("ir.PoolInfoProperties")
class PoolInfoProperties(Object):
    """PoolInfo object holds information related to memory pools
    where the statically sized allocate nodes will pooled into.

    Parameters
    ----------
    size_hint_bytes : Optional[int]
        The expected size hint to be used by the allocator.
        The default value would be -1 which means the pool
        is not size restricted.

    clock_frequency_hz : Optional[int]
        The clock frequency that the memory pool runs at in Hz.
        If not specified/known, this will default to -1 indicating
        it hasn't been defined.

    read_bandwidth_bytes_per_cycle : Optional[int]
        The read bandwidth of the memory pool in bytes/cycle.
        If not specified/known, this will default to -1 indicating
        it hasn't been defined.

    write_bandwidth_bytes_per_cycle : Optional[int]
        The write bandwidth of the memory pool in bytes/cycle.
        If not specified/known, this will default to -1 indicating
        it hasn't been defined.

    read_latency_cycles : Optional[int]
        The read latency of the memory pool in cycles.
        If not specified/known, this will default to 0.

    write_latency_cycles : Optional[int]
        The write latency of the memory pool in cycles.
        If not specified/known, this will default to 0.

    target_burst_bytes : Optional[Union[Dict[Target, int], None]]
        The burst length of the memory pool in bytes per target.
        If not specified/known for a given target, a burst length
        of 1 byte will be assumed.

    """

    def __init__(
        self,
        size_hint_bytes: Optional[int] = -1,
        clock_frequency_hz: Optional[int] = -1,
        read_bandwidth_bytes_per_cycle: Optional[int] = -1,
        write_bandwidth_bytes_per_cycle: Optional[int] = -1,
        read_latency_cycles: Optional[int] = 0,
        write_latency_cycles: Optional[int] = 0,
        target_burst_bytes=None,
    ):
        if not target_burst_bytes:
            target_burst_bytes = dict()

        self.__init_handle_by_constructor__(
            _ffi_api.PoolInfoProperties,  # type: ignore # pylint: disable=no-member
            size_hint_bytes,
            clock_frequency_hz,
            read_bandwidth_bytes_per_cycle,
            write_bandwidth_bytes_per_cycle,
            read_latency_cycles,
            write_latency_cycles,
            target_burst_bytes,
        )


@register_object("ir.ConstantInfo")
class ConstantInfo(Object):
    """ConstantInfo object hold information on a constant pool.

    Parameters
    ----------
    name_hint : str
        Name of the constant.
    byte_offset : int
        The byte_offset of the constant.
    data : NDArray
        The data of the constant.
    """

    def __init__(
        self,
        name_hint: str,
        byte_offset: int,
        data: NDArray,
    ):
        self.__init_handle_by_constructor__(
            _ffi_api.ConstantInfo,  # type: ignore # pylint: disable=no-member
            name_hint,
            byte_offset,
            data,
        )


@register_object("ir.WorkspacePoolInfo")
class WorkspacePoolInfo(PoolInfo):
    """WorkspacePoolInfo object holds information related to RW memory pools
    where the statically sized allocate nodes will pooled into.

    Parameters
    ----------
    pool_name : str
        The name of the memory pool

    targets : list[Target]
        A list of targets which could access the pool

    pool_info_properties : PoolInfoProperties
        The properties of the pool.
    """

    def __init__(
        self,
        pool_name: str,
        targets,
        pool_info_properties=None,
    ):
        super().__init__()

        if pool_info_properties is None:
            pool_info_properties = PoolInfoProperties()

        self.__init_handle_by_constructor__(
            _ffi_api.WorkspacePoolInfo,  # type: ignore # pylint: disable=no-member
            pool_name,
            targets,
            pool_info_properties,
        )


@register_object("ir.ConstantPoolInfo")
class ConstantPoolInfo(PoolInfo):
    """ConstantPoolInfo object holds information related to RO memory pools
    where the statically sized allocate nodes are pooled into.

    Parameters
    ----------
    pool_name : str
        The name of the memory pool

    targets : list[Target]
        describes which targets could access the pool

    pool_info_properties : PoolInfoProperties
        The properties of the pool.
    """

    def __init__(
        self,
        pool_name: str,
        targets,  # list[Target]
        constant_info_arr=None,  # list[ConstantInfo]
        pool_info_properties=None,
    ):
        super().__init__()

        if constant_info_arr is None:
            constant_info_arr = []
        if pool_info_properties is None:
            pool_info_properties = PoolInfoProperties()
        self.__init_handle_by_constructor__(
            _ffi_api.ConstantPoolInfo,  # type: ignore # pylint: disable=no-member
            pool_name,
            targets,
            constant_info_arr,
            pool_info_properties,
        )


@register_object("ir.WorkspaceMemoryPools")
class WorkspaceMemoryPools(Object):
    """This object contains a list of WorkspacePoolInfo objects to be used as
    workspace memory in the compilation

    Parameters
    ----------
    pools : List[WorkspacePoolInfo]
        The list of ConstantPoolInfo objects to be used with the compilation
    """

    def __init__(
        self,
        pools: List[WorkspacePoolInfo],
    ):
        self.__init_handle_by_constructor__(
            _ffi_api.WorkspaceMemoryPools, pools  # type: ignore # pylint: disable=no-member
        )


@register_object("ir.ConstantMemoryPools")
class ConstantMemoryPools(Object):
    """This object contains a list of ConstantPoolInfo objects to be used as
    read-only memory in the compilation

    Parameters
    ----------
    pools : List[ConstantPoolInfo]
        The list of ConstantPoolInfo objects to be used with the compilation
    """

    def __init__(
        self,
        pools: List[ConstantPoolInfo],
    ):
        self.__init_handle_by_constructor__(
            _ffi_api.ConstantMemoryPools, pools  # type: ignore # pylint: disable=no-member
        )


@register_object("ir.AllocatedPoolInfo")
class AllocatedPoolInfo(Object):
    """Allocate memory in a given pool.

    Parameters
    ----------
    pool : PoolInfo
        The pool in which to allocate memory.
    allocated_size : int
        The size of memory to allocate.
    """

    def __init__(
        self,
        pool: PoolInfo,
        allocated_size: int,
        pool_var_idx: int = 0,
    ):
        self.__init_handle_by_constructor__(
            _ffi_api.AllocatedPoolInfo, pool, allocated_size, pool_var_idx  # type: ignore # pylint: disable=no-member
        )

File: src/ir/memory_pools.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file src/ir/memory_pools.cc
 * \brief The object definition for relay.build argument type of memory pools
 */

#include <tvm/ir/memory_pools.h>
#include <tvm/relay/executor.h>

namespace tvm {

PoolInfo::PoolInfo(String pool_name, Integer size_hint_bytes, Integer clock_frequency_hz,
                   Integer read_bandwidth_bytes_per_cycle, Integer write_bandwidth_bytes_per_cycle,
                   Integer read_latency_cycles, Integer write_latency_cycles,
                   Map<Target, Integer> target_burst_bytes, Bool is_internal) {
  auto poolinfo_node = make_object<PoolInfoNode>();
  poolinfo_node->pool_name = pool_name;
  poolinfo_node->size_hint_bytes = size_hint_bytes;
  poolinfo_node->clock_frequency_hz = clock_frequency_hz;
  poolinfo_node->read_bandwidth_bytes_per_cycle = read_bandwidth_bytes_per_cycle;
  poolinfo_node->write_bandwidth_bytes_per_cycle = write_bandwidth_bytes_per_cycle;
  poolinfo_node->read_latency_cycles = read_latency_cycles;
  poolinfo_node->write_latency_cycles = write_latency_cycles;
  poolinfo_node->target_burst_bytes = target_burst_bytes;
  poolinfo_node->is_internal = is_internal;
  data_ = std::move(poolinfo_node);
}

TVM_REGISTER_NODE_TYPE(PoolInfoNode);

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<PoolInfoNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const PoolInfoNode*>(ref.get());
      p->stream << "PoolInfoNode(\n"
                << "  pool_name=" << node->pool_name
                << ",\n  size_hint_bytes=" << node->size_hint_bytes
                << ",\n  clock_frequency_hz=" << node->clock_frequency_hz
                << ",\n  read_bandwidth_bytes_per_cycle=" << node->read_bandwidth_bytes_per_cycle
                << ",\n  write_bandwidth_bytes_per_cycle=" << node->write_bandwidth_bytes_per_cycle
                << ",\n  read_latency_cycles=" << node->read_latency_cycles
                << ",\n  write_latency_cycles=" << node->write_latency_cycles
                << ",\n  target_burst_bytes=" << node->target_burst_bytes << ")";
    });

PoolInfoProperties::PoolInfoProperties(Integer size_hint_bytes, Integer clock_frequency_hz,
                                       Integer read_bandwidth_bytes_per_cycle,
                                       Integer write_bandwidth_bytes_per_cycle,
                                       Integer read_latency_cycles, Integer write_latency_cycles,
                                       Map<Target, Integer> target_burst_bytes, Bool is_internal) {
  auto poolinfo_properties_node = make_object<PoolInfoPropertiesNode>();
  poolinfo_properties_node->size_hint_bytes = size_hint_bytes;
  poolinfo_properties_node->clock_frequency_hz = clock_frequency_hz;
  poolinfo_properties_node->read_bandwidth_bytes_per_cycle = read_bandwidth_bytes_per_cycle;
  poolinfo_properties_node->write_bandwidth_bytes_per_cycle = write_bandwidth_bytes_per_cycle;
  poolinfo_properties_node->read_latency_cycles = read_latency_cycles;
  poolinfo_properties_node->write_latency_cycles = write_latency_cycles;
  poolinfo_properties_node->target_burst_bytes = target_burst_bytes;
  poolinfo_properties_node->is_internal = is_internal;
  data_ = std::move(poolinfo_properties_node);
}

TVM_REGISTER_NODE_TYPE(PoolInfoPropertiesNode);
TVM_REGISTER_GLOBAL("ir.PoolInfoProperties")
    .set_body_typed([](Integer size_hint_bytes, Integer clock_frequency_hz,
                       Integer read_bandwidth_bytes_per_cycle,
                       Integer write_bandwidth_bytes_per_cycle, Integer read_latency_cycles,
                       Integer write_latency_cycles, Map<Target, Integer> target_burst_bytes) {
      return PoolInfoProperties(size_hint_bytes, clock_frequency_hz, read_bandwidth_bytes_per_cycle,
                                write_bandwidth_bytes_per_cycle, read_latency_cycles,
                                write_latency_cycles, target_burst_bytes, Bool(false));
    });

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<PoolInfoPropertiesNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const PoolInfoPropertiesNode*>(ref.get());
      p->stream << "PoolInfoPropertiesNode(\n"
                << "  size_hint_bytes=" << node->size_hint_bytes
                << ",\n  clock_frequency_hz=" << node->clock_frequency_hz
                << ",\n  read_bandwidth_bytes_per_cycle=" << node->read_bandwidth_bytes_per_cycle
                << ",\n  write_bandwidth_bytes_per_cycle=" << node->write_bandwidth_bytes_per_cycle
                << ",\n  read_latency_cycles=" << node->read_latency_cycles
                << ",\n  write_latency_cycles=" << node->write_latency_cycles
                << ",\n  target_burst_bytes=" << node->target_burst_bytes << ")";
    });

WorkspacePoolInfo::WorkspacePoolInfo(String pool_name, Array<Target> targets,
                                     PoolInfoProperties properties) {
  auto poolinfo_node = make_object<WorkspacePoolInfoNode>();
  poolinfo_node->pool_name = pool_name;
  poolinfo_node->size_hint_bytes = properties->size_hint_bytes;
  poolinfo_node->targets = targets;
  poolinfo_node->clock_frequency_hz = properties->clock_frequency_hz;
  poolinfo_node->read_bandwidth_bytes_per_cycle = properties->read_bandwidth_bytes_per_cycle;
  poolinfo_node->write_bandwidth_bytes_per_cycle = properties->write_bandwidth_bytes_per_cycle;
  poolinfo_node->read_latency_cycles = properties->read_latency_cycles;
  poolinfo_node->write_latency_cycles = properties->write_latency_cycles;
  poolinfo_node->target_burst_bytes = properties->target_burst_bytes;
  poolinfo_node->is_internal = properties->is_internal;
  data_ = std::move(poolinfo_node);
}

TVM_REGISTER_NODE_TYPE(WorkspacePoolInfoNode);
TVM_REGISTER_GLOBAL("ir.WorkspacePoolInfo")
    .set_body_typed([](String pool_name, Array<Target> targets, PoolInfoProperties properties) {
      return WorkspacePoolInfo(pool_name, targets, properties);
    });

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<WorkspacePoolInfoNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const WorkspacePoolInfoNode*>(ref.get());
      p->stream << "WorkspacePoolInfoNode(\n"
                << "  pool_name=" << node->pool_name << ",\n  targets=" << node->targets
                << ",\n  size_hint_bytes=" << node->size_hint_bytes
                << ",\n  clock_frequency_hz=" << node->clock_frequency_hz
                << ",\n  read_bandwidth_bytes_per_cycle=" << node->read_bandwidth_bytes_per_cycle
                << ",\n  write_bandwidth_bytes_per_cycle=" << node->write_bandwidth_bytes_per_cycle
                << ",\n  read_latency_cycles=" << node->read_latency_cycles
                << ",\n  write_latency_cycles=" << node->write_latency_cycles
                << ",\n  target_burst_bytes=" << node->target_burst_bytes
                << ",\n  is_internal=" << node->is_internal << ")"
                << "\n";
    });

ConstantInfo::ConstantInfo(String name_hint, Integer byte_offset, runtime::NDArray data) {
  auto constant_info_node = make_object<ConstantInfoNode>();
  constant_info_node->name_hint = name_hint;
  constant_info_node->byte_offset = byte_offset;
  constant_info_node->data = data;
  data_ = std::move(constant_info_node);
}

TVM_REGISTER_NODE_TYPE(ConstantInfoNode);
TVM_REGISTER_GLOBAL("ir.ConstantInfo")
    .set_body_typed([](String name_hint, Integer byte_offset, runtime::NDArray data) {
      return ConstantInfo(name_hint, byte_offset, data);
    });
TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<ConstantInfoNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const ConstantInfoNode*>(ref.get());
      p->stream << "ConstantInfoNode(\n"
                << "name_hint=" << node->name_hint << ",\n byte_offset=" << node->byte_offset
                << ",\n data=" << node->data << ")";
    });

ConstantPoolInfo::ConstantPoolInfo(String pool_name, Array<Target> targets,
                                   Array<ConstantInfo> constant_info_array,
                                   PoolInfoProperties properties) {
  auto constant_poolinfo_node = make_object<ConstantPoolInfoNode>();
  constant_poolinfo_node->pool_name = pool_name;
  constant_poolinfo_node->constant_info_array = constant_info_array;
  constant_poolinfo_node->targets = targets;

  constant_poolinfo_node->size_hint_bytes = properties->size_hint_bytes;
  constant_poolinfo_node->clock_frequency_hz = properties->clock_frequency_hz;
  constant_poolinfo_node->read_bandwidth_bytes_per_cycle =
      properties->read_bandwidth_bytes_per_cycle;
  constant_poolinfo_node->write_bandwidth_bytes_per_cycle =
      properties->write_bandwidth_bytes_per_cycle;
  constant_poolinfo_node->read_latency_cycles = properties->read_latency_cycles;
  constant_poolinfo_node->write_latency_cycles = properties->write_latency_cycles;
  constant_poolinfo_node->target_burst_bytes = properties->target_burst_bytes;
  constant_poolinfo_node->is_internal = properties->is_internal;
  data_ = std::move(constant_poolinfo_node);
}

TVM_REGISTER_NODE_TYPE(ConstantPoolInfoNode);
TVM_REGISTER_GLOBAL("ir.ConstantPoolInfo")
    .set_body_typed([](String pool_name, Array<Target> targets,
                       Array<ConstantInfo> constant_info_array, PoolInfoProperties properties) {
      return ConstantPoolInfo(pool_name, targets, constant_info_array, properties);
    });

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<ConstantPoolInfoNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const ConstantPoolInfoNode*>(ref.get());
      p->stream << "ConstantPoolInfoNode(\n"
                << "  pool_name=" << node->pool_name << ",\n  targets=" << node->targets
                << ",\n  constant_info_array=" << node->constant_info_array
                << ",\n  size_hint_bytes=" << node->size_hint_bytes
                << ",\n  clock_frequency_hz=" << node->clock_frequency_hz
                << ",\n  read_bandwidth_bytes_per_cycle=" << node->read_bandwidth_bytes_per_cycle
                << ",\n  write_bandwidth_bytes_per_cycle=" << node->write_bandwidth_bytes_per_cycle
                << ",\n  read_latency_cycles=" << node->read_latency_cycles
                << ",\n  write_latency_cycles=" << node->write_latency_cycles
                << ",\n  target_burst_bytes=" << node->target_burst_bytes
                << ",\n  is_internal=" << node->is_internal << ")"
                << "\n";
    });

WorkspaceMemoryPools::WorkspaceMemoryPools(Array<PoolInfo> pools) {
  auto workspace_memory_pools_node = make_object<WorkspaceMemoryPoolsNode>();
  workspace_memory_pools_node->pools = pools;
  data_ = std::move(workspace_memory_pools_node);
}

TVM_REGISTER_NODE_TYPE(WorkspaceMemoryPoolsNode);
TVM_REGISTER_GLOBAL("ir.WorkspaceMemoryPools").set_body_typed([](Array<PoolInfo> pools) {
  return WorkspaceMemoryPools(pools);
});

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<WorkspaceMemoryPoolsNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const WorkspaceMemoryPoolsNode*>(ref.get());
      p->stream << "WorkspaceMemoryPoolsNode(\n"
                << "pools=" << node->pools << ")";
    });

ConstantMemoryPools::ConstantMemoryPools(Array<ConstantPoolInfo> pools) {
  auto constant_memory_pools_node = make_object<ConstantMemoryPoolsNode>();
  constant_memory_pools_node->pools = pools;
  data_ = std::move(constant_memory_pools_node);
}

TVM_REGISTER_NODE_TYPE(ConstantMemoryPoolsNode);
TVM_REGISTER_GLOBAL("ir.ConstantMemoryPools").set_body_typed([](Array<ConstantPoolInfo> pools) {
  return ConstantMemoryPools(pools);
});

TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
    .set_dispatch<ConstantMemoryPoolsNode>([](const ObjectRef& ref, ReprPrinter* p) {
      auto* node = static_cast<const ConstantMemoryPoolsNode*>(ref.get());
      p->stream << "ConstantMemoryPoolsNode(\n"
                << "pools=" << node->pools << ")";
    });
}  // namespace tvm
